src/hal/drivers/hal_bb_gpio.c

```diff
@@ -78,10 +78,33 @@ void configure_control_module() {
 }
 
 void configure_gpio_port(int n) {
+    volatile void *cm_per;  // pointer to clock manager registers
+    volatile unsigned int *regptr;
+    unsigned int regvalue;
+
     int fd = rtapi_open_as_root("/dev/mem", O_RDWR);
 
     gpio_ports[n] = hal_malloc(sizeof(bb_gpio_port));
 
+    // need to verify that port is enabled and clocked before accessing it
+    // port 0 is always mapped, the others need checked
+    if ( n > 0 ) {
+	cm_per = mmap(0,CM_PER_LEN, PROT_READ | PROT_WRITE, MAP_SHARED, fd, CM_PER_ADDR);
+	if(cm_per == MAP_FAILED) {
+	    rtapi_print_msg(RTAPI_MSG_ERR, "%s: ERROR: Unable to map Clock Module: %s\n", modname, strerror(errno));
+	    exit(1);
```

Comment:
Oops.

The rules are definitely changing.  Pre-uspace the whole open(/dev/mem)
thing would have been illegal.  What is the proper way to exit?


On Sat, Aug 20, 2016, at 06:23 PM, Jeff Epler wrote:
> >      int fd = rtapi_open_as_root("/dev/mem", O_RDWR);
> >  
> >      gpio_ports[n] = hal_malloc(sizeof(bb_gpio_port));
> >  
> > +    // need to verify that port is enabled and clocked before accessing it
> > +    // port 0 is always mapped, the others need checked
> > +    if ( n > 0 ) {
> > +	cm_per = mmap(0,CM_PER_LEN, PROT_READ | PROT_WRITE, MAP_SHARED, fd, CM_PER_ADDR);
> > +	if(cm_per == MAP_FAILED) {
> > +	    rtapi_print_msg(RTAPI_MSG_ERR, "%s: ERROR: Unable to map Clock Module: %s\n", modname, strerror(errno));
> > +	    exit(1);
> 
> It looks like you were merely following the surrounding code style by calling exit here.  I don't know if this is OK in MK, but in linuxcnc it isn't; calling exit() here would make the whole rtapi_app process exit.  looks like it'll be just about as easy to take this patch first and then go back and fix it, though.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/LinuxCNC/linuxcnc/pull/155/files/11252fc297b5720692e08380e20225c95ca10145#r75585709


-- 
  John Kasunich
  jmkasunich@fastmail.fm